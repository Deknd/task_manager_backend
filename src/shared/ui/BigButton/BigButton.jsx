

import React, { useState, useEffect } from 'react';
import style from "./bigButton.module.css"
import {
	scale_onClick_Active,
	scale_onHover_Active,
	scale_AtRest,
	transformation_Time,
	margin_OnClick,
	margin_atRest,
    widthShadow,
    opacityShadow,
    colorShdowRBG,
    textWhenNoFunction,
    noDescription,
    noColorButtonRBG,
 } from './bigButtonConstants';
import { Link, useLocation } from 'react-router-dom';


export const BigButton = (props) => {

    const {
        toRoute,
        description,
        img,
        onClickSent,
        colorButtonRBG,
    } =props;

    

    //ХУк для проверки, активена ли данная кнопка, если да, то выделяет ее тенью
    const location = useLocation();
    useEffect(() => {
        const currentPath = location.pathname;
        if (currentPath.endsWith(`/${toRoute}`)) {
            setSelect(true)
          } else setSelect(false) 
	 }, [location]);



    //Хук для слежения выбран данный компонент или нет
    const [isSelect, setSelect] = useState(false);
    const descriptionButton = description ? description : noDescription
    const colorButtonRBGSet = colorButtonRBG ? colorButtonRBG : noColorButtonRBG;




    //Метод обработки нажатия кнопки
    const shouldDoOnClick =!onClickSent && !toRoute ? ()=>{console.log(textWhenNoFunction,description)} : onClickSent

    //Хуки для слежения взаимодействия с кнопкой
    const [isHovered, setIsHovered] = useState(false);
    const [isActive, setIsActive] = useState(false);

  const handleMouseEnter = () => {
    setIsHovered(true);
  };

  const handleMouseLeave = () => {
    setIsHovered(false);
    setIsActive(false); 
  };

  const handleMouseDown = () => {
    setIsActive(true);
  };

  const handleMouseUp = () => {
    setIsActive(false);
  };



  //Изменяемый стиль, для кнопки
  const imageStyle = {
    transform: isActive
     ? `scale(${scale_onClick_Active})`
      : isHovered
       ? `scale(${scale_onHover_Active})`
        : `scale(${scale_AtRest})`,
    transition: `transform ${transformation_Time}s ease-in-out`,
    paddingTop: isActive ? `${margin_OnClick}px` : `${margin_atRest}`,
    boxShadow: isSelect ? `0 0 ${widthShadow}px rgba(${colorShdowRBG}, ${opacityShadow})` : 'none',
    backgroundColor: `rgba(${colorButtonRBGSet})`,
  };



return (
    <Link 
    to={toRoute}
    className={style.big_button} 
    role="button"
    onClick={shouldDoOnClick} 
    onMouseEnter={handleMouseEnter} 
    onMouseLeave={handleMouseLeave} 
    onMouseDown={handleMouseDown} 
    onMouseUp={handleMouseUp}
    style={imageStyle}>
        <div className={style.img_big_button}>
             {img ? <img className={style.image} src={img} alt={descriptionButton} /> : null}
        </div>
        <div className={style.text_big_button}>
            <span>
             {descriptionButton}
            </span>
             
        </div>
    </Link>
    );
}