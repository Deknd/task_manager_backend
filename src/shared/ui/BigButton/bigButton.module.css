.big_button {
    width: 100%;
    display: flex;
    justify-content: flex-end;
    align-items: center;
    height: 3rem;
    border-radius: 15px;
    text-decoration: none;
    color: black;
    
  }
  .big_button:hover{
    color: black;
  }
.img_big_button{
    width: 4rem;
   
    
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.image{
    width: 2rem;
    height: 2rem;
    

}
.text_big_button {
    margin-left: 0.7rem;
    flex-grow: 1;
    flex-shrink: 1;
    font-size: 1.6rem;
  }
