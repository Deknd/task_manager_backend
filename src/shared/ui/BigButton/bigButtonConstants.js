


//При клике изображение уменьшается до этого значения
export const scale_onClick_Active = 0.99; 
//Когда наводят мышку на изображение, оно увеличивается до этого значения
export const scale_onHover_Active = 1.03;
//Размер изображения в покое
export const scale_AtRest = 1;
//Время перехода между изображениями
export const transformation_Time = 0.1;
//при нажатии на изображение, оно смещается вниз на данное количество пикселей
export const margin_OnClick = 2;
//положение изображение в покое
export const margin_atRest = 0;
//толщина тени в пикселях
export const widthShadow = '8';
//прозрачность тени в процентах
export const opacityShadow = '0.3';
//цвет тени в rbg 
export const colorShdowRBG = '0, 0, 0';
//если не передан в кнопку роут и функция, то пишется в консоль это сообщение
export const textWhenNoFunction = 'Не передана ни одно действие на кнопку: ';
//если нет описание, то будет отображаться на кнопке
export const noDescription = 'Нет описания';
//если не передается wцвет кнопки, то кнока будет такого цвета
export const noColorButtonRBG = '0, 255, 0, 0.65';