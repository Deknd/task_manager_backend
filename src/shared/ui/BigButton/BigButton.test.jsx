import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from "react-dom";
import '@testing-library/jest-dom/extend-expect';

import iconTest from './test.png';

import { BigButton } from './BigButton';
import {
	scale_onClick_Active,
	scale_onHover_Active,
	scale_AtRest,
	transformation_Time,
	margin_OnClick,
	margin_atRest,
    widthShadow,
    opacityShadow,
    colorShdowRBG,
    textWhenNoFunction,
    noColorButtonRBG,
    noDescription,
 } from './bigButtonConstants';
import { MemoryRouter } from 'react-router-dom';


 describe('BigButton', () => {
    let container = null;
    beforeEach(() => {
        container = document.createElement("div");
        document.body.appendChild(container);
    });
    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
      });
     
    it('Проверка, как рисуется кнопка со всеми данными', () => {
        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';


        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} description={testDescription} img={iconTest} onClickSent={click} colorButtonRBG={testColor} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.getByAltText(testDescription);
        const description = screen.getByText(testDescription);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(1);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${testColor})`);
        expect(image.getAttribute('src')).toBe('test.png');
        expect(description.textContent).toBe(testDescription);

    })
    it('Проверка, как рисуется кнопка без функции', () => {
        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';


        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} description={testDescription} img={iconTest} colorButtonRBG={testColor} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.getByAltText(testDescription);
        const description = screen.getByText(testDescription);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(0);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${testColor})`);
        expect(image.getAttribute('src')).toBe('test.png');
        expect(description.textContent).toBe(testDescription);
    })

    it('Проверка, как рисуется кнопка без иконки', () => {
        const testRoute = '/test/route';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';
        const testDescription = 'test description';



        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} description={testDescription} colorButtonRBG={testColor} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.queryByAltText(testDescription);
        const description = screen.getByText(testDescription);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(0);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${testColor})`);
        expect(image).toBeNull();
        expect(description.textContent).toBe(testDescription);
    })

    it('Проверка, как рисуется кнопка без оисания', () => {
        const testRoute = '/test/route';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';
        const testDescription = 'test description';


        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} colorButtonRBG={testColor} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.queryByAltText(testDescription);
        const description = screen.queryByText(testDescription);
        const noDescriptionTest = screen.queryByText(noDescription);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(0);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${testColor})`);
        expect(noDescriptionTest).not.toBeNull();
        expect(noDescriptionTest.textContent).toBe(noDescription);

        expect(image).toBeNull();
        expect(description).toBeNull();
    })
    it('Проверка, как рисуется кнопка без оисания', () => {
        const testRoute = '/test/route';
        const click = jest.fn();
        const testDescription = 'test description';


        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.queryByAltText(testDescription);
        const description = screen.queryByText(testDescription);
        const noDescriptionTest = screen.queryByText(noDescription);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(0);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${noColorButtonRBG})`);
        expect(noDescriptionTest).not.toBeNull();
        expect(noDescriptionTest.textContent).toBe(noDescription);

        expect(image).toBeNull();
        expect(description).toBeNull();
    })

    it('Проверка, как рисуется анимация кнопки', () => {
        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';


        render(
        (
        <MemoryRouter>
            <BigButton toRoute={testRoute} description={testDescription} img={iconTest} onClickSent={click} colorButtonRBG={testColor} />
        </MemoryRouter>
        ), container )

        const button = screen.getByRole('button');
        const link = button.closest('a');
        const image = screen.getByAltText(testDescription);
        const description = screen.getByText(testDescription);


        expect(link.style.transform).toContain(`scale(${scale_AtRest})`);
        expect(link.style.paddingTop).toContain(`${margin_atRest}px`);


        fireEvent.mouseEnter(link);

        expect(link.style.transform).toContain(`scale(${scale_onHover_Active})`);
        expect(link.style.paddingTop).toContain(`${margin_atRest}px`);

        fireEvent.mouseDown(link);

        expect(link.style.transform).toContain(`scale(${scale_onClick_Active})`);
        expect(link.style.paddingTop).toContain(`${margin_OnClick}px`);

        fireEvent.mouseUp(link);

        expect(link.style.transform).toContain(`scale(${scale_onHover_Active})`);
        expect(link.style.paddingTop).toContain(`${margin_atRest}px`);

        fireEvent.mouseLeave(link);

        expect(link.style.transform).toContain(`scale(${scale_AtRest})`);
        expect(link.style.paddingTop).toContain(`${margin_atRest}px`);

        fireEvent.click(link);



        expect(click).toHaveBeenCalledTimes(1);

        expect(link.getAttribute('href')).toBe(`${testRoute}`);
        expect(link.getAttribute('style')).toContain(`background-color: rgba(${testColor})`);
        expect(image.getAttribute('src')).toBe('test.png');
        expect(description.textContent).toBe(testDescription);

    })

    it('BigButton snapshot', () => {

        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';
    
        const button =  render(
            (
            <MemoryRouter>
                <BigButton toRoute={testRoute} description={testDescription} img={iconTest} onClickSent={click} colorButtonRBG={testColor} />
            </MemoryRouter>
            ), container )
    
        expect(button).toMatchSnapshot();
    
      });

      it('BigButton snapshot мышка наведена', () => {

        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';
    
        const button =  render(
            (
            <MemoryRouter>
                <BigButton toRoute={testRoute} description={testDescription} img={iconTest} onClickSent={click} colorButtonRBG={testColor} />
            </MemoryRouter>
            ), container );

        const buttonTest = screen.getByRole('button');
        fireEvent.mouseEnter(buttonTest);

    
        expect(button).toMatchSnapshot();
    
      });

      it('BigButton snapshot кнопка нажата', () => {

        const testRoute = '/test/route';
        const testDescription = 'test description';
        const click = jest.fn();
        const testColor = '100, 50, 30, 0.65';
    
        const button =  render(
            (
            <MemoryRouter>
                <BigButton toRoute={testRoute} description={testDescription} img={iconTest} onClickSent={click} colorButtonRBG={testColor} />
            </MemoryRouter>
            ), container );

        const buttonTest = screen.getByRole('button');
        fireEvent.mouseDown(buttonTest);

    
        expect(button).toMatchSnapshot();
    
      });
 })