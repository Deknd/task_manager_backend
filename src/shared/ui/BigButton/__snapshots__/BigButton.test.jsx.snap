// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BigButton BigButton snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <a
        class="big_button"
        href="/test/route"
        role="button"
        style="transform: scale(1); transition: transform 0.1s ease-in-out; padding-top: 0px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
      >
        <div
          class="img_big_button"
        >
          <img
            alt="test description"
            class="image"
            src="test.png"
          />
        </div>
        <div
          class="text_big_button"
        >
          <span>
            test description
          </span>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="big_button"
      href="/test/route"
      role="button"
      style="transform: scale(1); transition: transform 0.1s ease-in-out; padding-top: 0px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
    >
      <div
        class="img_big_button"
      >
        <img
          alt="test description"
          class="image"
          src="test.png"
        />
      </div>
      <div
        class="text_big_button"
      >
        <span>
          test description
        </span>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BigButton BigButton snapshot кнопка нажата 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <a
        class="big_button"
        href="/test/route"
        role="button"
        style="transform: scale(0.99); transition: transform 0.1s ease-in-out; padding-top: 2px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
      >
        <div
          class="img_big_button"
        >
          <img
            alt="test description"
            class="image"
            src="test.png"
          />
        </div>
        <div
          class="text_big_button"
        >
          <span>
            test description
          </span>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="big_button"
      href="/test/route"
      role="button"
      style="transform: scale(0.99); transition: transform 0.1s ease-in-out; padding-top: 2px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
    >
      <div
        class="img_big_button"
      >
        <img
          alt="test description"
          class="image"
          src="test.png"
        />
      </div>
      <div
        class="text_big_button"
      >
        <span>
          test description
        </span>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`BigButton BigButton snapshot мышка наведена 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <a
        class="big_button"
        href="/test/route"
        role="button"
        style="transform: scale(1.03); transition: transform 0.1s ease-in-out; padding-top: 0px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
      >
        <div
          class="img_big_button"
        >
          <img
            alt="test description"
            class="image"
            src="test.png"
          />
        </div>
        <div
          class="text_big_button"
        >
          <span>
            test description
          </span>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="big_button"
      href="/test/route"
      role="button"
      style="transform: scale(1.03); transition: transform 0.1s ease-in-out; padding-top: 0px; box-shadow: none; background-color: rgba(100, 50, 30, 0.65);"
    >
      <div
        class="img_big_button"
      >
        <img
          alt="test description"
          class="image"
          src="test.png"
        />
      </div>
      <div
        class="text_big_button"
      >
        <span>
          test description
        </span>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
